engine/verification/fetcher/engine_test.go

```diff
@@ -1 +1,838 @@
+	}
+}
+
+// mockRequester mocks the chunk data pack requester with the given chunk data pack requests.
+// Each chunk should be requested exactly once.
+// On reply, it invokes the handler function with the given collection and chunk data pack for the chunk ID.
+func mockRequester(t *testing.T, requester *mockfetcher.ChunkDataPackRequester,
+	requests map[flow.Identifier]*verification.ChunkDataPackRequest,
+	chunkDataPacks map[flow.Identifier]*flow.ChunkDataPack,
+	collections map[flow.Identifier]*flow.Collection,
+	allExecutors flow.IdentityList, handler func(flow.Identifier, *flow.ChunkDataPack, *flow.Collection)) *sync.WaitGroup {
+
+	mu := sync.Mutex{}
+	wg := &sync.WaitGroup{}
+	wg.Add(len(requests))
+	requester.On("Request", mock.Anything, mock.Anything).Run(func(args mock.Arguments) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		actualRequest, ok := args[0].(*verification.ChunkDataPackRequest)
+		require.True(t, ok)
+
+		expectedRequest, ok := requests[actualRequest.ChunkID]
+		require.True(t, ok, "requester received an unexpected chunk request")
+
+		require.Equal(t, *expectedRequest, *actualRequest)
+
+		actualExecutors, ok := args[1].(flow.IdentityList)
+		require.True(t, ok)
+		require.ElementsMatchf(t, allExecutors, actualExecutors, "execution nodes lists do not match")
+
+		go func() {
+			cdp, ok := chunkDataPacks[actualRequest.ChunkID]
+			require.True(t, ok)
+
+			collection, ok := collections[actualRequest.ChunkID]
+			require.True(t, ok)
+
+			handler(actualRequest.Agrees[0], cdp, collection)
+			wg.Done()
+		}()
+	}).Return().Times(len(requests))
+
+	return wg
+}
+
+// chunkDataPackResponseFixture creates chunk data pack and collections for given chunks.
+func chunkDataPackResponseFixture(chunks flow.ChunkList) (map[flow.Identifier]*flow.ChunkDataPack,
+	map[flow.Identifier]*flow.Collection) {
+	chunkDataPacks := make(map[flow.Identifier]*flow.ChunkDataPack)
+	collections := make(map[flow.Identifier]*flow.Collection)
+
+	for _, chunk := range chunks {
+		chunkID := chunk.ID()
+		collection := unittest.CollectionFixture(1)
+		collections[chunkID] = &collection
+		chunkDataPacks[chunkID] = unittest.ChunkDataPackFixture(chunkID,
+			unittest.WithStartState(chunk.StartState),
+			unittest.WithCollectionID(collection.ID()))
+	}
+
+	return chunkDataPacks, collections
+}
+
+// verifiableChunkFixture is a test helper that creates verifiable chunks, chunk data packs,
+// and collection fixtures for the given chunks list.
+func verifiableChunkFixture(chunks flow.ChunkList, block *flow.Block, result *flow.ExecutionResult) (
+	map[flow.Identifier]*flow.ChunkDataPack,
+	map[flow.Identifier]*flow.Collection,
+	map[flow.Identifier]*verification.VerifiableChunkData) {
+
+	chunkDataPacks, collections := chunkDataPackResponseFixture(chunks)
+
+	verifiableChunks := make(map[flow.Identifier]*verification.VerifiableChunkData)
+	for _, chunk := range chunks {
+		c := chunk // avoids shallow copy of loop variable
```

Comment:
chunks is a `ChunkList` (`type ChunkList []*Chunk`), the loop variable `chunk` is hence a pointer variable. It has a fixed address in memory within the loop while iterating on the `Chunk` pointers in `chunks`. If we directly store this loop variable into another placeholder, at the end of the loop, we have the address of the last element in the list in all of the placeholders. Since all of them point to the same memory address of the loop variable. 

By creating a new variable each time we account for a new pointer variable per `chunk`, so each placeholder points to a distinct `chunk`. We experienced this issue a lot during the development of the networking layer.

Reply:
`chunk` is a loop variable whose _value_ is the address of a chunk. The address of the variable called `chunk` is fixed for the scope of the loop, so if we were taking a reference of `chunk`, or were referencing chunk outside the control flow of the loop (ie. in a goroutine), we would have the problem you're referring to. However we are copying the value of `chunk` within the control flow of the loop (on line 791) and are not using a reference to the loop variable. 

Here's an example with `int` pointers: https://play.golang.org/p/3OtUZZqLXUw.

Reply:
Oh sorry you are right, I missed that we are taking its _value_ not the _address_, fixed in: https://github.com/onflow/flow-go/pull/633/commits/2dcbcbf15517ebde27982dba361bfaa37d9092af